stablehlo/transforms/StablehloLegalizeDeprecatedOps.cpp

```diff
@@ -0,0 +1,162 @@
+};
+
+///////////
+// Patterns
+///////////
+
+struct CrossReplicaSumToAllReducePattern
+    : public OpRewritePattern<CrossReplicaSumOp> {
+  using OpRewritePattern<CrossReplicaSumOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(CrossReplicaSumOp crossReplicaSumOp,
+                                PatternRewriter &rewriter) const override {
+    auto allReduceOp = rewriter.replaceOpWithNewOp<AllReduceOp>(
+        crossReplicaSumOp, crossReplicaSumOp.getType(),
+        crossReplicaSumOp.getOperand(), crossReplicaSumOp.getReplicaGroups(),
+        /*channel_handle=*/ChannelHandleAttr(),
+        /*use_global_device_ids=*/false);
+
+    auto *block = rewriter.createBlock(&allReduceOp.getComputation());
+    auto elementType =
+        RankedTensorType::get({}, allReduceOp.getType().getElementType());
+    auto location = allReduceOp.getComputation().getLoc();
+    block->addArguments({elementType, elementType}, {location, location});
+    auto addOp = rewriter.create<AddOp>(location, block->getArgument(0),
+                                        block->getArgument(1));
+    rewriter.create<ReturnOp>(location, addOp.getResult());
+    return success();
+  }
+};
+
+///////////////////////
+// DRR helper functions
+///////////////////////
+
+DenseElementsAttr getScalarOfType(Type ty, int64_t rawValue) {
+  RankedTensorType scalarTy = RankedTensorType::get({}, ty);
+
+  if (auto floatTy = mlir::dyn_cast<FloatType>(ty)) {
+    APFloat value(floatTy.getFloatSemantics(), rawValue);
+    return DenseElementsAttr::get(scalarTy, value);
+  }
+  if (auto intTy = mlir::dyn_cast<IntegerType>(ty)) {
+    APInt value(intTy.getWidth(), static_cast<int64_t>(rawValue),
+                /*isSigned=*/true);
+    return DenseElementsAttr::get(scalarTy, value);
+  }
+  if (auto complexTy = mlir::dyn_cast<ComplexType>(ty)) {
+    if (auto floatTy = mlir::cast<FloatType>(complexTy.getElementType())) {
+      APFloat real(floatTy.getFloatSemantics(), rawValue);
+      APFloat imag = APFloat::getZero(floatTy.getFloatSemantics());
+      return DenseElementsAttr::get(scalarTy,
+                                    std::complex<APFloat>(real, imag));
+    }
+  }
+  llvm_unreachable("unsupported type");
+}
+
+DotDimensionNumbersAttr getDefaultDotDimensionNumbers(mlir::Value dotOpLhs) {
+  return DotDimensionNumbersAttr::get(
+      dotOpLhs.getContext(),
+      /*lhsBatchingDimensions=*/{},
+      /*rhsBatchingDimensions=*/{},
+      /*lhsContractingDimensions=*/
+      {cast<ShapedType>(dotOpLhs.getType()).getRank() - 1},
+      /*rhsContractingDimensions=*/{0});
+}
+
+DenseI64ArrayAttr getBroadcastDimensions(RankedTensorType resultType,
+                                         DenseI64ArrayAttr broadcastSizes) {
+  int64_t operandRank = resultType.getRank() - broadcastSizes.size();
+  SmallVector<int64_t> broadcastDimensions;
+  for (int64_t i = 0; i < operandRank; ++i) {
+    broadcastDimensions.push_back(i + broadcastSizes.size());
+  }
```

Comment:
Nit: this can also be written as

```
auto broadcastDimensions = llvm::map_to_vector(llvm::seq(operandRank), [](int64_t i) { return i + broadcastSizes.size(); });
```